### ISSUE TITLE
Would it be a good idea to make the api unable to panic?

### ISSUE THREAD

**Issue by JSorngard** [DomiDre/gauss-quad#56]
Right now any user of the crate can not decide if they want improper input to panic or not. This makes writing code that calls the functions much more concise and nice to write, but at the expense of taking choice away from the users.

If we want to ensure at compile time that the user of the crate inputs types with correct restrictions we could use the [newtype Idiom](https://doc.rust-lang.org/rust-by-example/generics/new_types.html). We would define types like
```rust
#[derive(...)]
pub struct Degree(usize);
impl Degree {
    pub const fn new(value: usize) -> Option<Self> {
        if value >= 2 {
            Some(Self(value))
        } else {
            None
        }
    }
}
impl TryFrom<usize> for Degree {...}
impl From<Degree> for usize {...}

// similarly for a type that represents `alpha` and `beta` in jacobi or laguerre.
pub struct QuadratureParameter(f64);
// trait impls
```
and then the `new` constructors of the quadrature structs would take those.

**Comment by JSorngard** [comment]
I probably prefer the second solution if we were to do this.

**Comment by DomiDre** [comment]
I agree it makes sense to give the user the possibility to handle panics. The latter choice where a Quad Rule returns a Result looks also better to my uneducated eye at least. As we are changing the API anyway, I see no problem and we should do it :+1:

**Comment by JSorngard** [comment]
I'll create a branch for experimenting with that.

**Comment by JSorngard** [comment]
I have implemented constructors that don't panic in the 56-no_panic_api branch, but I strongly dislike the implementation I managed. Perhaps you have a better idea for how to do it?
